File: stripe/src/test/java/com/stripe/android/PaymentSessionTest.java
package com.stripe.android;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;

import androidx.test.core.app.ApplicationProvider;

import com.stripe.android.exception.APIConnectionException;
import com.stripe.android.exception.APIException;
import com.stripe.android.exception.AuthenticationException;
import com.stripe.android.exception.CardException;
import com.stripe.android.exception.InvalidRequestException;
import com.stripe.android.model.Customer;
import com.stripe.android.model.PaymentMethod;
import com.stripe.android.model.PaymentMethodCreateParams;
import com.stripe.android.model.PaymentMethodFixtures;
import com.stripe.android.model.PaymentMethodTest;
import com.stripe.android.testharness.TestEphemeralKeyProvider;
import com.stripe.android.view.ActivityStarter;
import com.stripe.android.view.PaymentFlowActivity;
import com.stripe.android.view.PaymentFlowActivityStarter;
import com.stripe.android.view.PaymentMethodsActivity;
import com.stripe.android.view.PaymentMethodsActivityStarter;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatchers;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.robolectric.RobolectricTestRunner;

import java.util.Objects;
import java.util.Set;
import java.util.concurrent.ThreadPoolExecutor;

import static android.app.Activity.RESULT_CANCELED;
import static android.app.Activity.RESULT_OK;
import static com.stripe.android.CustomerSessionTest.FIRST_SAMPLE_KEY_RAW;
import static com.stripe.android.CustomerSessionTest.FIRST_TEST_CUSTOMER_OBJECT;
import static com.stripe.android.CustomerSessionTest.SECOND_TEST_CUSTOMER_OBJECT;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Test class for {@link PaymentSession}
 */
@RunWith(RobolectricTestRunner.class)
public class PaymentSessionTest {

    @NonNull private static final Customer FIRST_CUSTOMER =
            Objects.requireNonNull(Customer.fromString(FIRST_TEST_CUSTOMER_OBJECT));
    @NonNull private static final Customer SECOND_CUSTOMER =
            Objects.requireNonNull(Customer.fromString(SECOND_TEST_CUSTOMER_OBJECT));

    @NonNull private final TestEphemeralKeyProvider mEphemeralKeyProvider =
            new TestEphemeralKeyProvider();

    @NonNull private final PaymentSessionData mPaymentSessionData = new PaymentSessionData();

    @Mock private Activity mActivity;
    @Mock private StripeApiHandler mApiHandler;
    @Mock private ThreadPoolExecutor mThreadPoolExecutor;
    @Mock private PaymentSession.PaymentSessionListener mPaymentSessionListener;
    @Mock private CustomerSession mCustomerSession;
    @Mock private ActivityStarter<PaymentMethodsActivity, PaymentMethodsActivityStarter.Args>
            mPaymentMethodsActivityStarter;
    @Mock private ActivityStarter<PaymentFlowActivity, PaymentFlowActivityStarter.Args>
            mPaymentFlowActivityStarter;
    @Mock private PaymentSessionPrefs mPaymentSessionPrefs;

    @Captor private ArgumentCaptor<PaymentSessionData> mPaymentSessionDataArgumentCaptor;
    @Captor private ArgumentCaptor<Intent> mIntentArgumentCaptor;

    @Before
    public void setup()
            throws CardException, APIException, InvalidRequestException, AuthenticationException,
            APIConnectionException {
        MockitoAnnotations.initMocks(this);

        PaymentConfiguration.init(ApiKeyFixtures.FAKE_PUBLISHABLE_KEY);

        assertNotNull(FIRST_CUSTOMER);

        final PaymentMethod paymentMethod =
                PaymentMethod.fromString(PaymentMethodTest.PM_CARD_JSON);
        assertNotNull(paymentMethod);

        when(mApiHandler.retrieveCustomer(anyString(), ArgumentMatchers.<ApiRequest.Options>any()))
                .thenReturn(FIRST_CUSTOMER, SECOND_CUSTOMER);
        when(mApiHandler.createPaymentMethod(
                ArgumentMatchers.<PaymentMethodCreateParams>any(),
                ArgumentMatchers.<ApiRequest.Options>any()
        )).thenReturn(paymentMethod);
        when(mApiHandler.setDefaultCustomerSource(
                anyString(),
                anyString(),
                ArgumentMatchers.<String>anyList(),
                anyString(),
                anyString(),
                ArgumentMatchers.<ApiRequest.Options>any()))
                .thenReturn(SECOND_CUSTOMER);

        doAnswer(new Answer() {
            @Override
            public Object answer(InvocationOnMock invocation) {
                invocation.<Runnable>getArgument(0).run();
                return null;
            }
        }).when(mThreadPoolExecutor).execute(any(Runnable.class));
    }

    @Test
    public void init_addsPaymentSessionToken_andFetchesCustomer() {
        final CustomerSession customerSession = createCustomerSession();
        CustomerSession.setInstance(customerSession);

        final PaymentSession paymentSession = new PaymentSession(mActivity);
        paymentSession.init(mPaymentSessionListener, new PaymentSessionConfig.Builder().build());

        assertTrue(customerSession.getProductUsageTokens()
                .contains(PaymentSession.TOKEN_PAYMENT_SESSION));

        verify(mPaymentSessionListener).onCommunicatingStateChanged(eq(true));
    }

    @Test
    public void init_whenEphemeralKeyProviderContinues_fetchesCustomerAndNotifiesListener() {
        mEphemeralKeyProvider.setNextRawEphemeralKey(FIRST_SAMPLE_KEY_RAW);
        CustomerSession.setInstance(createCustomerSession());

        final PaymentSession paymentSession = new PaymentSession(mActivity);
        paymentSession.init(mPaymentSessionListener, new PaymentSessionConfig.Builder().build());
        verify(mPaymentSessionListener).onCommunicatingStateChanged(eq(true));
        verify(mPaymentSessionListener).onPaymentSessionDataChanged(any(PaymentSessionData.class));
        verify(mPaymentSessionListener).onCommunicatingStateChanged(eq(false));
    }

    @Test
    public void setCartTotal_setsExpectedValueAndNotifiesListener() {
        mEphemeralKeyProvider.setNextRawEphemeralKey(FIRST_SAMPLE_KEY_RAW);
        CustomerSession.setInstance(createCustomerSession());

        final PaymentSession paymentSession = new PaymentSession(mActivity);
        paymentSession.init(mPaymentSessionListener, new PaymentSessionConfig.Builder().build());
        paymentSession.setCartTotal(500L);

        verify(mPaymentSessionListener)
                .onPaymentSessionDataChanged(mPaymentSessionDataArgumentCaptor.capture());
        final PaymentSessionData data = mPaymentSessionDataArgumentCaptor.getValue();
        assertNotNull(data);
        assertEquals(500L, data.getCartTotal());
    }

    @Test
    public void handlePaymentData_whenPaymentMethodRequest_notifiesListenerAndFetchesCustomer() {
        CustomerSession.setInstance(createCustomerSession());

        PaymentSession paymentSession = new PaymentSession(mActivity);
        paymentSession.init(mPaymentSessionListener, new PaymentSessionConfig.Builder().build());

        // We have already tested the functionality up to here.
        reset(mPaymentSessionListener);

        final boolean handled = paymentSession.handlePaymentData(
                PaymentSession.PAYMENT_METHOD_REQUEST, RESULT_OK,
                new Intent().putExtra(PaymentMethodsActivity.EXTRA_SELECTED_PAYMENT,
                        PaymentMethod.fromString(PaymentMethodTest.PM_CARD_JSON)));
        assertTrue(handled);

        verify(mPaymentSessionListener)
                .onPaymentSessionDataChanged(mPaymentSessionDataArgumentCaptor.capture());
        final PaymentSessionData data = mPaymentSessionDataArgumentCaptor.getValue();
        assertNotNull(data);
        assertEquals(PaymentMethod.fromString(PaymentMethodTest.PM_CARD_JSON),
                data.getPaymentMethod());
    }

    @Test
    public void selectPaymentMethod_launchesPaymentMethodsActivityWithLog() {
        CustomerSession.setInstance(createCustomerSession());

        final PaymentSession paymentSession = new PaymentSession(mActivity);
        paymentSession.init(mPaymentSessionListener, new PaymentSessionConfig.Builder().build());

        paymentSession.presentPaymentMethodSelection();

        verify(mActivity).startActivityForResult(mIntentArgumentCaptor.capture(),
                eq(PaymentSession.PAYMENT_METHOD_REQUEST));

        final Intent intent = mIntentArgumentCaptor.getValue();
        assertNotNull(intent.getComponent());
        assertEquals(PaymentMethodsActivity.class.getName(),
                intent.getComponent().getClassName());

        final PaymentMethodsActivityStarter.Args args =
                PaymentMethodsActivityStarter.Args.create(intent);
        assertFalse(args.shouldRequirePostalCode);
    }

    @Test
    public void presentPaymentMethodSelection_withShouldRequirePostalCode_shouldPassInIntent() {
        CustomerSession.setInstance(createCustomerSession());

        final PaymentSession paymentSession = new PaymentSession(mActivity);
        paymentSession.init(mPaymentSessionListener, new PaymentSessionConfig.Builder().build());
        paymentSession.presentPaymentMethodSelection(true);

        verify(mActivity).startActivityForResult(mIntentArgumentCaptor.capture(),
                eq(PaymentSession.PAYMENT_METHOD_REQUEST));
        assertTrue(PaymentMethodsActivityStarter.Args.create(mIntentArgumentCaptor.getValue())
                .shouldRequirePostalCode);
    }

    @Test
    public void getSelectedPaymentMethodId_whenPrefsNotSet_returnsNull() {
        when(mCustomerSession.getCachedCustomer()).thenReturn(FIRST_CUSTOMER);
        CustomerSession.setInstance(mCustomerSession);
        assertNull(createPaymentSession().getSelectedPaymentMethodId(null));
    }

    @Test
    public void getSelectedPaymentMethodId_whenHasPaymentSessionData_returnsExpectedId() {
        mPaymentSessionData.setPaymentMethod(PaymentMethodFixtures.CARD_PAYMENT_METHOD);
        assertEquals("pm_123456789",
                createPaymentSession().getSelectedPaymentMethodId(null));
    }

    @Test
    public void getSelectedPaymentMethodId_whenHasPrefsSet_returnsExpectedId() {
        final String customerId = Objects.requireNonNull(FIRST_CUSTOMER.getId());
        when(mPaymentSessionPrefs.getSelectedPaymentMethodId(customerId)).thenReturn("pm_12345");

        when(mCustomerSession.getCachedCustomer()).thenReturn(FIRST_CUSTOMER);
        CustomerSession.setInstance(mCustomerSession);

        assertEquals("pm_12345",
                createPaymentSession().getSelectedPaymentMethodId(null));
    }

    @Test
    public void getSelectedPaymentMethodId_whenHasUserSpecifiedPaymentMethod_returnsExpectedId() {
        mPaymentSessionData.setPaymentMethod(PaymentMethodFixtures.CARD_PAYMENT_METHOD);
        assertEquals("pm_987",
                createPaymentSession().getSelectedPaymentMethodId("pm_987"));
    }

    @Test
    public void init_withoutSavedState_clearsLoggingTokensAndStartsWithPaymentSession() {
        final CustomerSession customerSession = createCustomerSession();
        CustomerSession.setInstance(customerSession);
        customerSession
                .addProductUsageTokenIfValid(PaymentMethodsActivity.TOKEN_PAYMENT_METHODS_ACTIVITY);
        assertEquals(1, customerSession.getProductUsageTokens().size());

        PaymentSession paymentSession = new PaymentSession(mActivity);
        paymentSession.init(mPaymentSessionListener, new PaymentSessionConfig.Builder().build());

        // The init removes PaymentMethodsActivity, but then adds PaymentSession
        final Set<String> loggingTokens = customerSession.getProductUsageTokens();
        assertEquals(1, loggingTokens.size());
        assertFalse(loggingTokens.contains(PaymentMethodsActivity.TOKEN_PAYMENT_METHODS_ACTIVITY));
        assertTrue(loggingTokens.contains(PaymentSession.TOKEN_PAYMENT_SESSION));
    }

    @Test
    public void init_withSavedStateBundle_doesNotClearLoggingTokens() {
        final CustomerSession customerSession = createCustomerSession();
        CustomerSession.setInstance(customerSession);
        customerSession
                .addProductUsageTokenIfValid(PaymentMethodsActivity.TOKEN_PAYMENT_METHODS_ACTIVITY);
        assertEquals(1, customerSession.getProductUsageTokens().size());

        PaymentSession paymentSession = new PaymentSession(mActivity);
        // If it is given any saved state at all, the tokens are not cleared out.
        paymentSession.init(mPaymentSessionListener,
                new PaymentSessionConfig.Builder().build(), new Bundle());

        final Set<String> loggingTokens = customerSession.getProductUsageTokens();
        assertEquals(2, loggingTokens.size());
        assertTrue(loggingTokens.contains(PaymentMethodsActivity.TOKEN_PAYMENT_METHODS_ACTIVITY));
        assertTrue(loggingTokens.contains(PaymentSession.TOKEN_PAYMENT_SESSION));
    }

    @Test
    public void completePayment_withLoggedActions_clearsLoggingTokensAndSetsResult() {
        final CustomerSession customerSession = createCustomerSession();
        CustomerSession.setInstance(customerSession);
        customerSession
                .addProductUsageTokenIfValid(PaymentMethodsActivity.TOKEN_PAYMENT_METHODS_ACTIVITY);
        assertEquals(1, customerSession.getProductUsageTokens().size());

        PaymentSession paymentSession = new PaymentSession(mActivity);
        // If it is given any saved state at all, the tokens are not cleared out.
        paymentSession.init(mPaymentSessionListener,
                new PaymentSessionConfig.Builder().build(), new Bundle());

        final Set<String> loggingTokens = customerSession.getProductUsageTokens();
        assertEquals(2, loggingTokens.size());

        reset(mPaymentSessionListener);

        paymentSession.onCompleted();
        assertTrue(customerSession.getProductUsageTokens().isEmpty());
    }

    @Test
    public void init_withSavedState_setsPaymentSessionData() {
        mEphemeralKeyProvider.setNextRawEphemeralKey(FIRST_SAMPLE_KEY_RAW);
        CustomerSession.setInstance(createCustomerSession());

        PaymentSession paymentSession = new PaymentSession(mActivity);
        paymentSession.init(mPaymentSessionListener, new PaymentSessionConfig.Builder().build());

        paymentSession.setCartTotal(300L);

        verify(mPaymentSessionListener)
                .onPaymentSessionDataChanged(mPaymentSessionDataArgumentCaptor.capture());
        final Bundle bundle = new Bundle();
        paymentSession.savePaymentSessionInstanceState(bundle);
        PaymentSessionData firstPaymentSessionData = mPaymentSessionDataArgumentCaptor.getValue();

        PaymentSession.PaymentSessionListener secondListener =
                mock(PaymentSession.PaymentSessionListener.class);

        paymentSession.init(secondListener, new PaymentSessionConfig.Builder().build(), bundle);
        verify(secondListener)
                .onPaymentSessionDataChanged(mPaymentSessionDataArgumentCaptor.capture());

        final PaymentSessionData secondPaymentSessionData =
                mPaymentSessionDataArgumentCaptor.getValue();
        assertEquals(firstPaymentSessionData.getCartTotal(),
                secondPaymentSessionData.getCartTotal());
        assertEquals(firstPaymentSessionData.getPaymentMethod(),
                secondPaymentSessionData.getPaymentMethod());
    }

    @Test
    public void handlePaymentData_withInvalidRequestCode_aborts() {
        final PaymentSession paymentSession = createPaymentSession();
        assertFalse(paymentSession.handlePaymentData(-1, RESULT_CANCELED, new Intent()));
        verify(mCustomerSession, never()).retrieveCurrentCustomer(
                ArgumentMatchers.<CustomerSession.CustomerRetrievalListener>any());
    }

    @Test
    public void handlePaymentData_withValidRequestCodeAndCanceledResult_retrievesCustomer() {
        final PaymentSession paymentSession = createPaymentSession();
        assertFalse(paymentSession.handlePaymentData(PaymentSession.PAYMENT_METHOD_REQUEST,
                RESULT_CANCELED, new Intent()));
        verify(mCustomerSession).retrieveCurrentCustomer(
                ArgumentMatchers.<CustomerSession.CustomerRetrievalListener>any());
    }

    @NonNull
    private PaymentSession createPaymentSession() {
        return new PaymentSession(mCustomerSession, mPaymentMethodsActivityStarter,
                mPaymentFlowActivityStarter, mPaymentSessionData,
                mPaymentSessionPrefs);
    }

    @NonNull
    private CustomerSession createCustomerSession() {
        return new CustomerSession(ApplicationProvider.getApplicationContext(),
                mEphemeralKeyProvider, null, mThreadPoolExecutor, mApiHandler,
                ApiKeyFixtures.FAKE_PUBLISHABLE_KEY,
                "acct_abc123", true);
    }
}

File: stripe/src/test/java/com/stripe/android/PaymentSessionTest.kt
package com.stripe.android

import android.app.Activity.RESULT_CANCELED
import android.app.Activity.RESULT_OK
import android.content.Context
import android.content.Intent
import androidx.activity.ComponentActivity
import androidx.lifecycle.Lifecycle
import androidx.test.core.app.ActivityScenario
import androidx.test.core.app.ApplicationProvider
import com.google.common.truth.Truth.assertThat
import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import com.stripe.android.model.Customer
import com.stripe.android.model.CustomerFixtures
import com.stripe.android.model.PaymentMethod
import com.stripe.android.model.PaymentMethodCreateParams
import com.stripe.android.model.PaymentMethodFixtures
import com.stripe.android.testharness.TestEphemeralKeyProvider
import com.stripe.android.view.ActivityScenarioFactory
import com.stripe.android.view.ActivityStarter
import com.stripe.android.view.BillingAddressFields
import com.stripe.android.view.PaymentFlowActivity
import com.stripe.android.view.PaymentFlowActivityStarter
import com.stripe.android.view.PaymentMethodsActivity
import com.stripe.android.view.PaymentMethodsActivityStarter
import java.util.concurrent.ThreadPoolExecutor
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue
import org.junit.runner.RunWith
import org.mockito.Mockito.doAnswer
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.reset
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows

/**
 * Test class for [PaymentSession]
 */
@RunWith(RobolectricTestRunner::class)
class PaymentSessionTest {

    private val ephemeralKeyProvider = TestEphemeralKeyProvider()

    private val threadPoolExecutor: ThreadPoolExecutor = mock()
    private val paymentSessionListener: PaymentSession.PaymentSessionListener = mock()
    private val customerSession: CustomerSession = mock()
    private val paymentMethodsActivityStarter:
        ActivityStarter<PaymentMethodsActivity, PaymentMethodsActivityStarter.Args> = mock()
    private val paymentFlowActivityStarter:
        ActivityStarter<PaymentFlowActivity, PaymentFlowActivityStarter.Args> = mock()

    private val paymentSessionDataArgumentCaptor: KArgumentCaptor<PaymentSessionData> = argumentCaptor()
    private val paymentMethodsActivityStarterArgsCaptor: KArgumentCaptor<PaymentMethodsActivityStarter.Args> = argumentCaptor()

    private val context: Context = ApplicationProvider.getApplicationContext()
    private val activityScenarioFactory = ActivityScenarioFactory(context)

    @BeforeTest
    fun setup() {
        PaymentConfiguration.init(context, ApiKeyFixtures.FAKE_PUBLISHABLE_KEY)

        doAnswer { invocation ->
            invocation.getArgument<Runnable>(0).run()
            null
        }.`when`<ThreadPoolExecutor>(threadPoolExecutor).execute(any())

        CustomerSession.instance = createCustomerSession()
    }

    @Test
    fun init_addsPaymentSessionToken_andFetchesCustomer() {
        createActivity {
            val paymentSession = PaymentSession(it, DEFAULT_CONFIG)
            paymentSession.init(paymentSessionListener)

            assertEquals(
                setOf(PaymentSession.TOKEN_PAYMENT_SESSION),
                CustomerSession.getInstance().productUsageTokens
            )

            verify(paymentSessionListener)
                .onCommunicatingStateChanged(eq(true))
        }
    }

    @Test
    fun init_whenEphemeralKeyProviderContinues_fetchesCustomerAndNotifiesListener() {
        ephemeralKeyProvider
            .setNextRawEphemeralKey(EphemeralKeyFixtures.FIRST_JSON)

        createActivity {
            val paymentSession = PaymentSession(it, DEFAULT_CONFIG)
            paymentSession.init(paymentSessionListener)
            verify(paymentSessionListener)
                .onCommunicatingStateChanged(true)
            verify(paymentSessionListener)
                .onCommunicatingStateChanged(false)
        }
    }

    @Test
    fun handlePaymentData_whenPaymentMethodSelected_notifiesListenerAndFetchesCustomer() {
        createActivity {
            val paymentSession = PaymentSession(it, DEFAULT_CONFIG)
            paymentSession.init(paymentSessionListener)

            // We have already tested the functionality up to here.
            reset(paymentSessionListener)

            val result = PaymentMethodsActivityStarter.Result(
                paymentMethod = PaymentMethodFixtures.CARD_PAYMENT_METHOD
            )
            val handled = paymentSession.handlePaymentData(
                PaymentMethodsActivityStarter.REQUEST_CODE, RESULT_OK,
                Intent().putExtras(result.toBundle()))
            assertTrue(handled)

            verify(paymentSessionListener)
                .onPaymentSessionDataChanged(paymentSessionDataArgumentCaptor.capture())
            val data = paymentSessionDataArgumentCaptor.firstValue
            assertEquals(PaymentMethodFixtures.CARD_PAYMENT_METHOD, data.paymentMethod)
            assertFalse(data.useGooglePay)
        }
    }

    @Test
    fun handlePaymentData_whenGooglePaySelected_notifiesListenerAndFetchesCustomer() {
        createActivity {
            val paymentSession = PaymentSession(it, DEFAULT_CONFIG)
            paymentSession.init(paymentSessionListener)

            // We have already tested the functionality up to here.
            reset(paymentSessionListener)

            val result = PaymentMethodsActivityStarter.Result(
                useGooglePay = true
            )
            val handled = paymentSession.handlePaymentData(
                PaymentMethodsActivityStarter.REQUEST_CODE, RESULT_OK,
                Intent().putExtras(result.toBundle()))
            assertTrue(handled)

            verify(paymentSessionListener)
                .onPaymentSessionDataChanged(paymentSessionDataArgumentCaptor.capture())
            val data = paymentSessionDataArgumentCaptor.firstValue
            assertNull(data.paymentMethod)
            assertTrue(data.useGooglePay)
        }
    }

    @Test
    fun selectPaymentMethod_launchesPaymentMethodsActivityWithLog() {
        createActivity { activity ->
            val paymentSession = PaymentSession(activity, DEFAULT_CONFIG)
            paymentSession.init(paymentSessionListener)
            paymentSession.presentPaymentMethodSelection()

            val nextStartedActivityForResult =
                Shadows.shadowOf(activity).nextStartedActivityForResult
            val intent = nextStartedActivityForResult.intent

            assertThat(nextStartedActivityForResult.requestCode)
                .isEqualTo(PaymentMethodsActivityStarter.REQUEST_CODE)
            assertThat(intent.component?.className)
                .isEqualTo(PaymentMethodsActivity::class.java.name)

            val args =
                PaymentMethodsActivityStarter.Args.create(intent)
            assertEquals(BillingAddressFields.Full, args.billingAddressFields)
        }
    }

    @Test
    fun presentPaymentMethodSelection_withShouldRequirePostalCode_shouldPassInIntent() {
        createActivity { activity ->
            val paymentSession = PaymentSession(
                activity,
                PaymentSessionConfig.Builder()
                    .setShippingMethodsRequired(false)
                    .setBillingAddressFields(BillingAddressFields.PostalCode)
                    .build()
            )
            paymentSession.init(paymentSessionListener)
            paymentSession.presentPaymentMethodSelection()

            val nextStartedActivityForResult =
                Shadows.shadowOf(activity).nextStartedActivityForResult
            val intent = nextStartedActivityForResult.intent

            assertThat(nextStartedActivityForResult.requestCode)
                .isEqualTo(PaymentMethodsActivityStarter.REQUEST_CODE)
            assertThat(intent.component?.className)
                .isEqualTo(PaymentMethodsActivity::class.java.name)

            val args =
                PaymentMethodsActivityStarter.Args.create(nextStartedActivityForResult.intent)
            assertThat(args.billingAddressFields)
                .isEqualTo(BillingAddressFields.PostalCode)
        }
    }

    @Test
    fun getSelectedPaymentMethodId_whenHasPaymentSessionData_returnsExpectedId() {
        createActivity {
            val paymentSession = createPaymentSession(
                it,
                DEFAULT_CONFIG,
                PaymentSessionFixtures.PAYMENT_SESSION_DATA.copy(
                    paymentMethod = PaymentMethodFixtures.CARD_PAYMENT_METHOD
                )
            )
            paymentSession.presentPaymentMethodSelection()
            verify(paymentMethodsActivityStarter).startForResult(
                paymentMethodsActivityStarterArgsCaptor.capture()
            )
            assertEquals(
                "pm_123456789",
                paymentMethodsActivityStarterArgsCaptor.firstValue.initialPaymentMethodId
            )
        }
    }

    @Test
    fun getSelectedPaymentMethodId_whenHasUserSpecifiedPaymentMethod_returnsExpectedId() {
        createActivity {
            val paymentSession = createPaymentSession(
                it,
                DEFAULT_CONFIG,
                PaymentSessionFixtures.PAYMENT_SESSION_DATA.copy(
                    paymentMethod = PaymentMethodFixtures.CARD_PAYMENT_METHOD
                )
            )
            paymentSession.presentPaymentMethodSelection("pm_987")
            verify(paymentMethodsActivityStarter).startForResult(
                paymentMethodsActivityStarterArgsCaptor.capture()
            )
            assertEquals(
                "pm_987",
                paymentMethodsActivityStarterArgsCaptor.firstValue.initialPaymentMethodId
            )
        }
    }

    @Test
    fun init_withoutSavedState_clearsLoggingTokensAndStartsWithPaymentSession() {
        val customerSession = CustomerSession.getInstance()
        customerSession
            .addProductUsageTokenIfValid(PaymentMethodsActivity.TOKEN_PAYMENT_METHODS_ACTIVITY)
        assertEquals(
            setOf(PaymentMethodsActivity.TOKEN_PAYMENT_METHODS_ACTIVITY),
            customerSession.productUsageTokens
        )

        createActivity {
            val paymentSession = PaymentSession(it, DEFAULT_CONFIG)
            paymentSession.init(paymentSessionListener)

            // The init removes PaymentMethodsActivity, but then adds PaymentSession
            assertEquals(
                setOf(PaymentSession.TOKEN_PAYMENT_SESSION),
                customerSession.productUsageTokens
            )
        }
    }

    @Test
    fun completePayment_withLoggedActions_clearsLoggingTokensAndSetsResult() {
        val customerSession = CustomerSession.getInstance()
        customerSession
            .addProductUsageTokenIfValid(PaymentMethodsActivity.TOKEN_PAYMENT_METHODS_ACTIVITY)
        assertEquals(
            setOf(PaymentMethodsActivity.TOKEN_PAYMENT_METHODS_ACTIVITY),
            customerSession.productUsageTokens
        )

        createActivity {
            val paymentSession = PaymentSession(it, DEFAULT_CONFIG)
            // If it is given any saved state at all, the tokens are not cleared out.
            paymentSession.init(paymentSessionListener)

            assertEquals(
                setOf(PaymentSession.TOKEN_PAYMENT_SESSION),
                customerSession.productUsageTokens
            )

            reset(paymentSessionListener)

            paymentSession.onCompleted()
            assertTrue(customerSession.productUsageTokens.isEmpty())
        }
    }

    @Test
    fun init_withSavedState_setsPaymentSessionData() {
        ephemeralKeyProvider.setNextRawEphemeralKey(EphemeralKeyFixtures.FIRST_JSON)

        createActivity { activity ->
            val paymentSession = PaymentSession(activity, DEFAULT_CONFIG).also {
                it.setCartTotal(300L)
                it.init(paymentSessionListener)
            }

            paymentSession.init(paymentSessionListener)

            verify(paymentSessionListener, times(2))
                .onPaymentSessionDataChanged(paymentSessionDataArgumentCaptor.capture())

            assertEquals(
                paymentSessionDataArgumentCaptor.allValues[0],
                paymentSessionDataArgumentCaptor.allValues[1]
            )
        }
    }

    @Test
    fun handlePaymentData_withInvalidRequestCode_aborts() {
        createActivity {
            val paymentSession = createPaymentSession(it)
            assertFalse(paymentSession.handlePaymentData(-1, RESULT_CANCELED, Intent()))
            verify(customerSession, never()).retrieveCurrentCustomer(any())
        }
    }

    @Test
    fun handlePaymentData_withPaymentMethodsActivityRequestCodeAndCanceledResult_doesNotRetrieveCustomer() {
        createActivity {
            val paymentSession = createPaymentSession(it)
            assertFalse(
                paymentSession.handlePaymentData(
                    PaymentMethodsActivityStarter.REQUEST_CODE,
                    RESULT_CANCELED,
                    Intent()
                )
            )
            verify(customerSession, never()).retrieveCurrentCustomer(any())
        }
    }

    @Test
    fun handlePaymentData_withPaymentFlowActivityRequestCodeAndCanceledResult_retrievesCustomer() {
        createActivity {
            val paymentSession = createPaymentSession(it)
            assertFalse(paymentSession.handlePaymentData(PaymentFlowActivityStarter.REQUEST_CODE,
                RESULT_CANCELED, Intent()))
            verify(customerSession).retrieveCurrentCustomer(any())
        }
    }

    @Test
    fun onDestroy_shouldReleaseListener() {
        createActivityScenario { activityScenario ->
            activityScenario.onActivity { activity ->
                val paymentSession = createPaymentSession(activity)
                paymentSession.init(paymentSessionListener)
                assertNotNull(paymentSession.listener)
                activityScenario.moveToState(Lifecycle.State.DESTROYED)
                assertNull(paymentSession.listener)
            }
        }
    }

    private fun createPaymentSession(
        activity: ComponentActivity,
        config: PaymentSessionConfig = DEFAULT_CONFIG,
        paymentSessionData: PaymentSessionData = PaymentSessionData(config)
    ): PaymentSession {
        return PaymentSession(
            activity,
            activity.application,
            activity,
            activity,
            activity,
            config,
            customerSession,
            paymentMethodsActivityStarter,
            paymentFlowActivityStarter,
            paymentSessionData
        )
    }

    private fun createCustomerSession(): CustomerSession {
        return CustomerSession(
            context,
            FakeStripeRepository(),
            ApiKeyFixtures.FAKE_PUBLISHABLE_KEY,
            "acct_abc123",
            threadPoolExecutor = threadPoolExecutor,
            ephemeralKeyManagerFactory = EphemeralKeyManager.Factory.Default(
                keyProvider = ephemeralKeyProvider,
                shouldPrefetchEphemeralKey = true
            )
        )
    }

    private fun createActivity(callback: (ComponentActivity) -> Unit) {
        // start an arbitrary ComponentActivity
        createActivityScenario { it.onActivity(callback) }
    }

    private fun createActivityScenario(
        callback: (ActivityScenario<out ComponentActivity>) -> Unit
    ) {
        activityScenarioFactory.create<PaymentMethodsActivity>(
            PaymentMethodsActivityStarter.Args.DEFAULT
        ).use(callback)
    }

    private class FakeStripeRepository : AbsFakeStripeRepository() {
        override fun createPaymentMethod(
            paymentMethodCreateParams: PaymentMethodCreateParams,
            options: ApiRequest.Options
        ): PaymentMethod {
            return PaymentMethodFixtures.CARD_PAYMENT_METHOD
        }

        override fun setDefaultCustomerSource(
            customerId: String,
            publishableKey: String,
            productUsageTokens: Set<String>,
            sourceId: String,
            sourceType: String,
            requestOptions: ApiRequest.Options
        ): Customer {
            return SECOND_CUSTOMER
        }

        override fun retrieveCustomer(
            customerId: String,
            requestOptions: ApiRequest.Options
        ): Customer {
            return FIRST_CUSTOMER
        }
    }

    private companion object {
        private val FIRST_CUSTOMER = CustomerFixtures.CUSTOMER
        private val SECOND_CUSTOMER = CustomerFixtures.OTHER_CUSTOMER

        private val DEFAULT_CONFIG = PaymentSessionFixtures.CONFIG
    }
}
